jetty-core/jetty-client/src/main/java/org/eclipse/jetty/client/transport/HttpConversation.java

```diff
@@ -32,115 +35,162 @@ public class HttpConversation extends Attributes.Lazy
     private final Deque<HttpExchange> exchanges = new ConcurrentLinkedDeque<>();
     private volatile ResponseListeners listeners;
 
-    public Deque<HttpExchange> getExchanges()
+    public void offerExchange(HttpExchange exchange)
     {
-        return exchanges;
+        exchanges.offer(exchange);
+    }
+
+    private HttpExchange firstExchange()
+    {
+        HttpExchange exchange = exchanges.peekFirst();
+        assert exchange != null;
+        return exchange;
+    }
+
+    public HttpExchange lastExchange()
+    {
+        HttpExchange exchange = exchanges.peekLast();
+        assert exchange != null;
```

Comment:
I would replace the assertion with a null check throwing `IllegalStateException` instead.